### PR TITLE
Changes default string to empty string

### LINKED ISSUE #86: opencdmsdb provider plugin
As an OpenCDMS core developer, I need an ORM implementation for OpenCDMS (e.g. `/opencdms/provider/opencdmsdb/`) so that I can use the OpenCDMS domain models (`/opencdms/models/`) with the default OpenCDMS database backend and not just with other supported CDMS databases.

### PR DESCRIPTION
**Pull request by chinedu117**
- Adds Optional[] type hints
- Changes default string to empty string 
-Tests passed

Closes #86

### CODE REVIEW

File: opencdms/models/cdm.py [hunk 1]
```diff
@@ -263,16 +263,16 @@ class Observation(OpenCDMSBase):
...
     valid_to: Optional[datetime] = field(default=None)
-    host_id: Optional[str] = field(default="")
-    observer_id: Optional[str] = field(default="")
+    host_id: Optional[int] = field(default=None)
+    observer_id: Optional[int] = field(default=None)
     observing_procedure_id: Optional[int] = field(default=None)
-    report_id: Optional[str] = field(default="")
-    collection_id: Optional[str] = field(default="")
-    feature_of_interest_id: Optional[str] = field(default="")
+    report_id: Optional[int] = field(default=None)
```
Comment (david-i-berry) [inline opencdms/models/cdm.py]: this needs to be a string and not optional.

File: opencdms/models/cdm.py [hunk 2]
```diff
@@ -263,16 +263,16 @@ class Observation(OpenCDMSBase):
...
     result_uom: Optional[str] = field(default="")
     result_description: Optional[str] = field(default="")
-    result_quality: Optional[dict] = field(default=None)
+    result_quality: Optional[str] = field(default=None)
     result_time: Optional[datetime] = field(default=None)
     valid_from: Optional[datetime] = field(default=None)
     valid_to: Optional[datetime] = field(default=None)
-    host_id: Optional[str] = field(default="")
-    observer_id: Optional[str] = field(default="")
+    host_id: Optional[int] = field(default=None)
```
Comment (david-i-berry) [inline opencdms/models/cdm.py]: str, not optional.

File: opencdms/models/cdm.py [hunk 3]
```diff
@@ -263,16 +263,16 @@ class Observation(OpenCDMSBase):
...
+    observer_id: Optional[int] = field(default=None)
```
Comment (david-i-berry) [inline opencdms/models/cdm.py]: string, not optional.

File: opencdms/models/cdm.py [hunk 4]
```diff
@@ -263,16 +263,16 @@ class Observation(OpenCDMSBase):
     phenomenon_start: Optional[datetime] = field(default=None)
     result_uom: Optional[str] = field(default="")
     result_description: Optional[str] = field(default="")
-    result_quality: Optional[dict] = field(default=None)
+    result_quality: Optional[str] = field(default=None)
```
Comment (david-i-berry) [inline opencdms/models/cdm.py]: dict or JSONB

File: opencdms/models/cdm.py [hunk 5]
```diff
@@ -263,16 +263,16 @@ class Observation(OpenCDMSBase):
...
-    host_id: Optional[str] = field(default="")
-    observer_id: Optional[str] = field(default="")
+    host_id: Optional[int] = field(default=None)
+    observer_id: Optional[int] = field(default=None)
     observing_procedure_id: Optional[int] = field(default=None)
-    report_id: Optional[str] = field(default="")
-    collection_id: Optional[str] = field(default="")
-    feature_of_interest_id: Optional[str] = field(default="")
+    report_id: Optional[int] = field(default=None)
+    collection_id: Optional[int] = field(default=None)
```
Comment (david-i-berry) [inline opencdms/models/cdm.py]: str

File: opencdms/models/cdm.py [hunk 6]
```diff
@@ -263,16 +263,16 @@ class Observation(OpenCDMSBase):
...
+    feature_of_interest_id: Optional[int] = field(default=None)
```
Comment (david-i-berry) [inline opencdms/models/cdm.py]: str

File: opencdms/models/cdm.py (hunk 2 above)
Comment (chinedu117) [reply opencdms/models/cdm.py]: Noted. Means we would also change the table schema.

File: opencdms/models/cdm.py (hunk 4 above)
Comment (david-i-berry) [reply opencdms/models/cdm.py]: Observation.id should also be a str and not optional.

**Review by david-i-berry — COMMENTED** [review]
Most of the IDs for the data elements (rather than code tables) will need to be UUIDs within the global climate data system. As such these need to be strings and not null / optional.

**Comment by chinedu117** [comment]
> Most of the IDs for the data elements (rather than code tables) will need to be UUIDs within the global climate data system. As such these need to be strings and not null / optional.

Thanks for the observation. Would it be safe to use UUID column type instead ? Then instead of str type we could use uuid type.

Also most of the foreignkeys would still be represented as optional so as to allow us to instantiate the object before attaching the necessary ids or objects. If we make the foreignkeys mandatory then we would not be able to create an instance of the class except the related object exists.

**Review by isedwards — APPROVED** [review]

**Pull request merged**